lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,245 @@
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in,out] function : LLVM function to be run on.
+bool PatchReadFirstLane::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Read-First-Lane\n");
+
+  m_divergenceAnalysis = &getAnalysis<LegacyDivergenceAnalysis>();
+
+  const bool changed = liftReadFirstLane(function);
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Specify what analysis passes this pass depends on.
+//
+// @param [in,out] analysisUsage : The place to record our analysis pass usage requirements.
+void PatchReadFirstLane::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
+  analysisUsage.addRequired<LegacyDivergenceAnalysis>();
+}
+
+// =====================================================================================================================
+// Try to lift readfirstlanes per block to transform SGPR instructions into VGPR instruction.
+// Reversely iterate over a block who has readfirstlane instructions. Try to find qualified instructions to progagate
+// readfirstlane to them. The qualified instruction is that all its users are proven to be uniform and the cost keeps
+// one readfirstlane inserted for a lift.
+//
+// @param [in,out] function : LLVM function to be run for readfirstlane optimization.
+bool PatchReadFirstLane::liftReadFirstLane(Function &function) {
+  // Collect initial amdgcn_readfirstlane and their corresponding blocks
+  DenseSet<BasicBlock *> readFirstLaneBlocks;
+  DenseSet<Instruction *> initialReadFirstLanes;
+  Module *module = function.getParent();
+  for (auto &func : *module) {
+    if (func.getIntrinsicID() == Intrinsic::amdgcn_readfirstlane) {
+      for (User *user : func.users()) {
+        Instruction *inst = cast<Instruction>(user);
+        if (inst->getFunction() != &function)
+          continue;
+        initialReadFirstLanes.insert(inst);
+        readFirstLaneBlocks.insert(inst->getParent());
+      }
+    }
+  }
+  bool changed = false;
+
+  // Try to lift readfirstlanes per block
+  for (auto curBb : readFirstLaneBlocks) {
+    // Collect the instructions that can be marked as uniform
+    DenseSet<Instruction *> markedUniforms(initialReadFirstLanes);
+    // Collect the instructions where the readfirstlanes can be lifted to
+    DenseSet<Instruction *> liftInsts;
+
+    // Reversely iterate over the block to find qualified instructions to lift readfirstlanes
+    for (BasicBlock::reverse_iterator instIt = curBb->rbegin(), endIt = curBb->rend(); instIt != endIt; ++instIt) {
+      Instruction *candidateInst = &*instIt;
+      // The candidate instrction for a lift should be marked as uniform
+      if (markedUniforms.find(candidateInst) != markedUniforms.end()) {
+        unsigned numOps = candidateInst->getNumOperands();
+        if (dyn_cast<CallInst>(candidateInst))
+          --numOps;
```

Comment:
Is this really correct? I don't think the caller is counted among `getNumOperands`, but even if it does, the count has to be consistent for iterating over the operands, right?

Reply:
- Yes, `getNumOperands` returns 2 for `call i32 @llvm.amdgcn.readfirstlane(i32 %.i35)`. 
- Yes, the count can be consistent because the return value has already marked as uniform.